src/utils/hooks/api/xcm/use-xcm-bridge.ts

```diff
@@ -1,90 +1,71 @@
 import { ApiProvider, Bridge, ChainName } from '@interlay/bridge/build';
-import { useEffect, useState } from 'react';
+import { useErrorHandler } from 'react-error-boundary';
+import { useQuery, UseQueryResult } from 'react-query';
 import { firstValueFrom } from 'rxjs';
 
 import { XCM_ADAPTERS } from '@/config/relay-chains';
 import { BITCOIN_NETWORK } from '@/constants';
 import { Chains } from '@/types/chains';
 
+import { getXCMEndpoints } from './get-xcm-endpoints';
+
+const XCMNetwork = BITCOIN_NETWORK === 'mainnet' ? 'mainnet' : 'testnet';
+
 const XCMBridge = new Bridge({
   adapters: Object.values(XCM_ADAPTERS)
 });
 
-// MEMO: BitcoinNetwork type is not available on XCM bridge
-const XCMNetwork = BITCOIN_NETWORK === 'mainnet' ? 'mainnet' : 'testnet';
+type XCMBridgeData = {
+  bridge: Bridge;
+  provider: ApiProvider;
+};
 
-const getDestinationChains = (chain: ChainName): Chains => XCMBridge.router.getDestinationChains(chain as any);
+type UseXCMBridge = UseQueryResult<XCMBridgeData | undefined> & {
+  getOriginatingChains: () => Chains;
+  getDestinationChains: (chain: ChainName) => Chains;
+  getAvailableTokens: (from: ChainName, to: ChainName) => string[];
+};
 
-const getAvailableTokens = (from: ChainName, to: ChainName): any => XCMBridge.router.getAvailableTokens({ from, to });
+const initXCMBridge = async () => {
+  const XCMProvider = new ApiProvider(XCMNetwork);
+  const chains = Object.keys(XCM_ADAPTERS) as ChainName[];
 
-// TODO: This config needs to be pushed higher up the app.
-// Not sure how this will look: something to decide when
-// adding USDT support.
-const getEndpoints = (chains: ChainName[]) => {
-  switch (true) {
-    case chains.includes('kusama'):
-      return {
-        kusama: ['wss://kusama-rpc.polkadot.io', 'wss://kusama.api.onfinality.io/public-ws'],
-        kintsugi: ['wss://api-kusama.interlay.io/parachain', 'wss://kintsugi.api.onfinality.io/public-ws'],
-        statemine: ['wss://statemine-rpc.polkadot.io', 'wss://statemine.api.onfinality.io/public-ws']
-      };
-    case chains.includes('polkadot'):
+  // TODO: Get rid of any casting - mismatch between ApiRx types
+  await firstValueFrom(XCMProvider.connectFromChain(chains, getXCMEndpoints(chains)) as any);
+
+  // Set Apis
+  await Promise.all(chains.map((chain: ChainName) => XCMBridge.findAdapter(chain).setApi(XCMProvider.getApi(chain))));
+
+  return { provider: XCMProvider, bridge: XCMBridge };
+};
+
+const useXCMBridge = (): UseXCMBridge => {
+  const queryKey = ['available-xcm-channels'];
+
+  const queryResult = useQuery({
+    queryKey,
+    queryFn: initXCMBridge
+  });
+
+  const { error } = queryResult;
+
+  const getOriginatingChains = (): Chains =>
```

Comment:
It's best to wrap these functions in `useCallback` cause it best for them to be memoized.